lib/assets/javascripts/turbograft/turbolinks.coffee

```diff
@@ -136,39 +136,48 @@ class window.Turbolinks
 
   @loadPage: (url, xhr, options = {}) ->
     triggerEvent 'page:receive'
+    response = new TurboGraft.Response(xhr)
+    options.onlyKeys ?= []
+    options.exceptKeys ?= []
     options.updatePushState ?= true
-    if upstreamDocument = new TurboGraft.Response(xhr).document()
-      if options.partialReplace
-        reflectNewUrl url if options.updatePushState
-        updateBody(upstreamDocument, xhr, options)
-      else
-        turbohead = new TurboGraft.TurboHead(activeDocument, upstreamDocument)
-        if turbohead.hasAssetConflicts()
-          return Turbolinks.fullPageNavigate(url)
-        reflectNewUrl url if options.updatePushState
-        turbohead.waitForAssets().then((result) ->
-          updateBody(upstreamDocument, xhr, options) unless result?.isCanceled
-        )
-    else
+    options.partialReplace = isPartialReplace(options, response)
+
+    unless upstreamDocument = response.document()
       triggerEvent 'page:error', xhr
       Turbolinks.fullPageNavigate(url)
+      return
+
+    if options.partialReplace
+      reflectNewUrl url if options.updatePushState
+      updateBody(upstreamDocument, response, options)
+      return
 
-  updateBody = (upstreamDocument, xhr, options) ->
+    turbohead = new TurboGraft.TurboHead(activeDocument, upstreamDocument)
+    if turbohead.hasAssetConflicts()
+      return Turbolinks.fullPageNavigate(url)
+
+    reflectNewUrl url if options.updatePushState
+    turbohead.waitForAssets().then((result) ->
+      updateBody(upstreamDocument, response, options) unless result?.isCanceled
+    )
+
+  updateBody = (upstreamDocument, response, options) ->
     nodes = changePage(
       upstreamDocument.querySelector('title')?.textContent,
       removeNoscriptTags(upstreamDocument.querySelector('body')),
       CSRFToken.get(upstreamDocument).token,
       'runScripts',
       options
     )
-    reflectRedirectedUrl(xhr) if options.updatePushState
-    options.onLoadFunction?()
+
+    reflectRedirectedUrl(response) if options.updatePushState
+    Turbolinks.resetScrollPosition() unless options.partialReplace
+
+    options.callback?()
     triggerEvent 'page:load', nodes
 
   changePage = (title, body, csrfToken, runScripts, options = {}) ->
     activeDocument.title = title if title
-    options.onlyKeys ?= []
-    options.exceptKeys ?= []
 
     if options.onlyKeys.length
```

Comment:
similarly, I think it might have been a historical mistake to have `options.onlyKeys ?= []` above, if only for this check to pass without exception -- `options.onlyKeys?.length`